packages/json-rpc/ws-provider/src/ws-provider.ts

```diff
@@ -1,56 +1,185 @@
+            (status = {
+              type: e ? WsEvent.ERROR : WsEvent.CLOSE,
+              event: e,
+            }),
+          )
+          setTimeout(reject, e ? 300 : 0, e)
+        }
+        const timeoutToken = setTimeout(() => {
+          const event = { type: "timeout" }
+          onStatusChanged(
+            (status = {
+              type: WsEvent.ERROR,
+              event,
+            }),
+          )
+          reject(event)
+        }, 1_000)
+        const cleanup = () => {
+          clearTimeout(timeoutToken)
+          socket.removeEventListener("error", onError)
           socket.removeEventListener("open", onOpen)
         }
-        socket.addEventListener("error", onError, { once: true })
+        socket.addEventListener("open", onOpen)
+        socket.addEventListener("error", onError)
+        disconnect = () => {
+          disconnect = () => onError(null as any)
+        }
       })
 
+      onStatusChanged(
+        (status = {
+          type: WsEvent.CONNECTED,
+          uri,
+          protocols,
+        }),
+      )
+
       return (onMessage, onHalt) => {
         const _onMessage = (e: MessageEvent) => {
           if (typeof e.data === "string") onMessage(e.data)
         }
-        const innerHalt = (reason: string) => () => {
-          console.warn(`WS halt (${reason})`)
-          onHalt()
-        }
-        const onError = innerHalt("err")
-        const onClose = innerHalt("close")
+        const innerHalt =
+          (reason: WsEvent.CLOSE | WsEvent.ERROR) => (e: any) => {
+            console.warn(`WS halt (${reason})`)
+            onStatusChanged(
+              (status = {
+                type: reason,
+                event: e,
+              }),
+            )
+            onHalt()
+          }
+        const onError = innerHalt(WsEvent.ERROR)
+        const onClose = innerHalt(WsEvent.CLOSE)
 
         socket.addEventListener("message", _onMessage)
         socket.addEventListener("error", onError)
         socket.addEventListener("close", onClose)
+        disconnect = (withHalt) => {
+          disconnect = () => {}
+          socket.removeEventListener("message", _onMessage)
+          socket.removeEventListener("error", onError)
+          socket.removeEventListener("close", onClose)
+          socket.close()
+          if (withHalt) onClose({})
+        }
 
         return {
           send: (msg) => {
             socket.send(msg)
           },
-          disconnect: () => {
-            socket.removeEventListener("message", _onMessage)
-            socket.removeEventListener("error", onError)
-            socket.removeEventListener("close", onClose)
-            socket.close()
-          },
+          disconnect,
         }
       }
-    })
+    }) as WsJsonRpcProvider
+
+    result.getStatus = () => status
+    result.switch = (...args) => {
+      if (status.type === WsEvent.CONNECTED) {
+        if (args.length) switchTo = args as any
+        disconnect(true)
+      }
```

Comment:
would it be possible to also switch if the status.type is connecting? Use case being that it's been trying to connect to an endpoint for more than 10s and for some reason it's not failing => I want to try another endpoint.

Reply:
It would complicate things a lot (b/c we don't have observables in here :/). However, the use-case that you mentioned is impossible because if it has been trying to stablish the connection for more than 1 sec, then it will automatically switch to the next one.

Reply:
I ended up making this change. I also tested it in the examle script.